util/src/error.rs

```diff
@@ -0,0 +1,165 @@
+use core::{
+    any::{Any, TypeId},
+    fmt,
+};
+
+/// `Error` is a trait representing the basic expectations for error values,
+/// i.e., values of type `E` in `Result<T, E>`.
+///
+/// This is essentially a re-implementation of the Rust standard library's
+/// `std::error::Error` trait. Because we cannot use `std` in the kernel, we
+/// must reimplement that trait in `mycelium_util`.
+///
+/// Errors must describe themselves through the `Display` and `Debug` traits,
+/// and may provide cause chain information:
+///
+/// The [`source`] method is generally used when errors cross "abstraction
+/// boundaries". If one module must report an error that is caused by an error
+/// from a lower-level module, it can allow access to that error via the
+/// [`source`] method. This makes it possible for the high-level module to
+/// provide its own errors while also revealing some of the implementation for
+/// debugging via [`source`] chains.
+///
+/// [`source`]: trait.Error.html#method.source
+pub trait Error: fmt::Debug + fmt::Display {
```

Comment:
Given that we don't need to perfectly match `std`, perhaps we should include a `'static` bound on `Error`, and avoid the need for such a bound on the other methods. IIRC `std` only doesn't have that bound for backwards-compatibility reasons.

Reply:
Hmm, not a bad idea! i was thinking about experimenting with some other APIs that std doesn’t have here, as well (error categories/severities?)

Reply:
Stuff like that would probably be feasible to add with non-defaulted impls in the future, so I don't think we need to worry about getting that right first try. Probably worth having some sort of error classification like that, though.